felix/bpf-gpl/conntrack_cleanup.c

```diff
@@ -24,237 +24,75 @@ struct ct_iter_ctx {
 	__u64 now;
 	__u64 end_time;
 
-	__u64 num_kvs_seen_normal;
-	__u64 num_kvs_seen_nat_fwd;
-	__u64 num_kvs_seen_nat_rev;
-	__u64 num_kvs_deleted_normal;
-	__u64 num_kvs_deleted_nat_fwd;
-	__u64 num_kvs_deleted_nat_rev;
+	__u64 num_cleaned;
 };
 
-// sub_age calculates now-then assuming that the difference is less than
-// 1<<63.  Values larger than that are assumed to have wrapped (then>now) and
-// 0 is returned in that case.
-static __u64 sub_age(__u64 now, __u64 then)
-{
-	__u64 age = now - then;
-	if (age > (1ull<<63)) {
-		// Wrapped, assume that means then > now.
-		return 0;
+static CALI_BPF_INLINE bool ct_key_equal(struct calico_ct_key *key1, struct calico_ct_key *key2) {
+	if (!key1 || !key2) {
+		return key1 == key2;
 	}
-	return age;
-}
-
-#define CT_VALUE_FINS_SEEN_DSR(value) (value->a_to_b.fin_seen || value->b_to_a.fin_seen)
-#define CT_VALUE_FINS_SEEN_NON_DSR(value) (value->a_to_b.fin_seen && value->b_to_a.fin_seen)
-#define CT_VALUE_ESTABLISHED(value) (value->a_to_b.syn_seen && value->a_to_b.ack_seen && value->b_to_a.syn_seen && value->b_to_a.ack_seen)
-#define CT_VALUE_DSR(value) (value->flags & CALI_CT_FLAG_DSR_FWD)
-
-// max_age returns the maximum age for the given conntrack "tracking" entry.
-static __u64 calculate_max_age(const struct calico_ct_key *key, const struct calico_ct_value *value)
-{
-	__u64 max_age;
-	switch (key->protocol) {
-	case IPPROTO_TCP:
-		if (value->a_to_b.rst_seen || value->b_to_a.rst_seen) {
-			max_age = __globals.tcp_reset_seen;
-		} else if (
-		    (CT_VALUE_DSR(value) && CT_VALUE_FINS_SEEN_DSR(value)) ||
-		    CT_VALUE_FINS_SEEN_NON_DSR(value)
-		) {
-			max_age = __globals.tcp_fins_seen;
-		} else if (CT_VALUE_ESTABLISHED(value) || CT_VALUE_DSR(value)) {
-			if (value->rst_seen) {
-				/* We have seen RST in the past, but we have seen traffic
-				 * since then so we want to be cautious and not tear down
-				 * the conntrack too soon in case the RST was spurious,
-				 * but we are also not sure if the connection is still
-				 * established.
-				 */
-				max_age = __globals.tcp_fins_seen;
-			} else {
-				max_age = __globals.tcp_established;
-			}
-		} else {
-			max_age = __globals.tcp_syn_sent;
-		}
-		break;
-	case IPPROTO_UDP:
-		max_age = __globals.udp_timeout;
-		break;
-	case IPPROTO_ICMP_46:
-		max_age = __globals.icmp_timeout;
-		break;
-	default:
-		max_age = __globals.generic_timeout;
-		break;
+	if ((key1->protocol == key2->protocol) &&
```

Comment:
could use strcmp? I think there is a compiler built in as well https://docs.ebpf.io/linux/helper-function/bpf_strncmp/

Reply:
```bpf_strncmp``` requires the 3rd param to be readonly.

Reply:
```__builtin_memcmp``` works fine.